### DISCUSSION TITLE
Uart1 pinmux bug

### DISCUSSION THREAD

**Discussion by Mat-Alm** [linux-chenxing/linux-chenxing.org#77] [General]
Hi! I'am trying to use the Uart1 adding it config to the .dts file like this:
I'am in the kernel version 6.1 rebase

&uart1 {
	pinctrl-names = "default";
	pinctrl-0 = <&uart1_mode2_pins>;
	status = "okay";

};
and alias

	aliases {
		serial0 = &pm_uart;
		serial1 = &uart1;
	};
Recompile and load the new kernel. But the system refuses load the driver, the log message that i get:

pinctrl-msc313 1f203c00.pinctrl: invalid group "uart1_mode1" for function "uart1" 

@fifteenhex  I had a look at the pinctrl-msc313.c and found the bug at:

static const char * const ssd20xd_uart1_groups[] = {
	GROUPNAME_UART1_MODE1
	GROUPNAME_UART1_MODE2
	GROUPNAME_UART1_MODE3
	GROUPNAME_UART1_MODE4
};

It need comas after each GRUOUPNAME fields 

static const char * const ssd20xd_uart1_groups[] = {
	GROUPNAME_UART1_MODE1,
	GROUPNAME_UART1_MODE2,
	GROUPNAME_UART1_MODE3,
	GROUPNAME_UART1_MODE4,
};

Now it works fine!

[    0.236048] 1f221000.serial: ttyS0 at MMIO 0x1f221000 (irq = 43, base_baud =   10800000) is a 16550A
[    0.970559] printk: console [ttyS0] enabled
[    0.975878] 1f221200.uart: ttyS1 at MMIO 0x1f221200 (irq = 44, base_baud = 10  800000) is a 16550A

**Comment by fifteenhex** [comment]
Hi,

Thanks for the report! I'll add this fix to my branch at some point.

**Reply by Mat-Alm** [reply]
Hi! This was just a fix for the driver loading. After this the uart does not receive and transmit nothing.
After further investigation i found out some problems in  mstar-v7.dtsi
At: buildroot/output/build/linux-mstar_v6_1_rebase/arch/arm
this fragments
				uart1_pins: uart1 {
					function = "uart1";
					groups = "fuart_cts_rts";
				};
				I don't  know if this is right:
                               uart1_cts_pins: uart1_cts {
					function = "uart1";
					groups = "fuart_cts";
				};

Was change to:
				uart1_pins: uart1 {
					function = "uart1";
					groups = "uart1";
				}; 

                             I don't  know if this is right:
				uart1_cts_pins: uart1_cts {
					function = "uart1";
					groups = "fuart_cts";
				};

And more changes at this: 

			uart1: uart@221200 {
				compatible = "mstar,msc313-uart", "snps,dw-apb-uart";
				reg = <0x221200 0x100>;
				reg-shift = <3>;
				interrupts-extended = <&intc_irq GIC_SPI 35 IRQ_TYPE_LEVEL_HIGH>;
				clocks = <&clkgen MSC313_CLKGEN_DEGLITCHES MSC313_CLKGEN_UART1 0>;
				status = "disabled";
			};

Was changed to:
			uart1: uart@221200 {
				compatible = "mstar,msc313-uart", "snps,dw-apb-uart";
				reg = <0x221200 0x100>;
				reg-shift = <3>;
				interrupts-extended = <&intc_irq GIC_SPI 35 IRQ_TYPE_LEVEL_HIGH>;
				clocks = <&clkgen MSC313_CLKGEN_DEGLITCHES MSC313_CLKGEN_UART1 0>;
				pinctrl-names = "default";
				pinctrl-0 = <&uart1_pins>;
				status = "disabled";
			};

Now driver load and loopback test works on uart1!

**Reply by fifteenhex** [reply]
Can you put your changes into a commit for me? That way I can pull your commit and integrate the changes without copy/paste mistakes :)

**Reply by Mat-Alm** [reply]
@fifteenhex  Done on 6.1 rebase kernel! 
I think it needs some more fix in some pin mux names for uart1.

Just some questions.

It's possible to add support to uart2? Or it's not available in the SD202 soc?

There is abetter method to upgrade/flash the system by uboot without eth or serial?
Seems uboot does not support usb or emmc/sdcard medias...

**Reply by fifteenhex** [reply]
Hi,

I don't think I really tested the other UARTs so it's possible there are a bunch of changes needed.
I think there is a uart2 on the SSD202D. I can't remember where it's wired up though.

But all of the UARTs should work as long as the pinctrl and clks are wired up. The Fast UART DMA thing isn't supported yet though.

**Reply by fifteenhex** [reply]
> There is abetter method to upgrade/flash the system by uboot without eth or serial?
> Seems uboot does not support usb or emmc/sdcard medias...

eMMC should work in u-boot but I think you'll need to enable it yourself.
The buildroot config I use has a "rescue" system that can boot and write the flash to avoid having to do it in u-boot though.

**Comment by Mat-Alm** [comment]
Hi! I make a merge request but still not reply
